src/Chainweb/Miner/POW.hs

```diff
@@ -132,120 +156,245 @@ powMiner logFun conf nid cutDb = runForever logFun "POW Miner" $ do
         --
         go gen (i + 1) (HM.filter (\(T2 h _) -> h > limit) adjustments')
 
-    -- | INVARIANT: A new cut, chain, and parent header is reselected after each
-    -- hash failure. This ensures that the Cut the miner is working on doesn't
-    -- grow stale, and cause forks. Without this condition (or a similar one),
-    -- the miners cause forks quite aggressively.
+awaitNextCut :: CutDb cas -> Cut -> IO Cut
+awaitNextCut cutDb c = atomically $ do
+    c' <- _cutStm cutDb
+    when (c' == c) retry
+    return c'
+
+mineCut
+    :: PayloadCas cas
+    => Given WebBlockHeaderDb
+    => Given (PayloadDb cas)
+    => LogFunction
+    -> MinerConfig
+    -> NodeId
+    -> CutDb cas
+    -> Cut
+    -> Nonce
+    -> Adjustments
+    -> IO (T3 BlockHeader Cut Adjustments)
+mineCut logfun conf nid cutDb !c !nonce !adjustments = do
+
+    -- Randomly pick a chain to mine on.
     --
-    mine
-        :: Given WebBlockHeaderDb
-        => Given (PayloadDb cas)
-        => Nonce
-        -> Adjustments
-        -> IO (T3 BlockHeader Cut Adjustments)
-    mine !nonce !adjustments = do
-        -- Get the current longest cut.
-        --
-        c <- _cut cutDb
+    cid <- randomChainId c
 
-        -- Randomly pick a chain to mine on.
-        --
-        cid <- randomChainId c
+    -- The parent block the mine on. Any given chain will always
+    -- contain at least a genesis block, so this otherwise naughty
+    -- `^?!` will always succeed.
+    --
+    let !p = c ^?! ixg cid
 
-        -- The parent block the mine on. Any given chain will always
-        -- contain at least a genesis block, so this otherwise naughty
-        -- `^?!` will always succeed.
-        --
-        let !p = c ^?! ixg cid
+    -- check if chain can be mined on (check adjacent parents)
+    --
+    case getAdjacentParents c p of
+
+        Nothing -> mineCut logfun conf nid cutDb c nonce adjustments
+            -- spin until a chain is found that isn't blocked
+
+        Just adjParents -> do
+
+            -- get payload
+            let mokPact = False
```

Comment:
This line can be removed by now.

Reply:
Yes, I think, for testing one should provide a mock `PactExcecutionService`.